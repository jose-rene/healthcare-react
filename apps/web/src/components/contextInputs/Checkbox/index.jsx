import React from "react";
import { useFormContext } from "../../../Context/FormContext";

const Checkbox = (
    { name, label = "", disableFn, wrapperClass, ...otherProps },
    ref
) => {
    const { form, getValue, onChange } = useFormContext();

    const id = name;
    const value = getValue ? getValue(name) : otherProps.value;

    const disabled = disableFn ? disableFn(form) : false;

    return (
        <div className={wrapperClass}>
            <input
                className="form-check-input m-1 me-2"
                type="checkbox"
                name={name}
                id={id}
                value={value}
                checked={disabled ? false : value}
                ref={ref}
                disabled={disabled}
                onChange={onChange}
            />
            <label className="form-check-label" htmlFor={name}>
                {label}
            </label>
        </div>
    );
};
Checkbox.displayName = "Checkbox";

export default React.forwardRef(Checkbox);
