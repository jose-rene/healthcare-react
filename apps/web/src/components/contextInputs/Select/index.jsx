import React, { useMemo } from "react";
import { Form, FloatingLabel } from "react-bootstrap";
import { useFormContext } from "Context/FormContext";

const Select = (
    {
        name,
        label = "",
        inlineLabel = false,
        options = [],
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        labelClass = "",
        wrapperClass = false,
        labelKey = "title",
        valueKey = "val",
        disabled = false,
        ...otherProps
    },
    ref
) => {
    const { getValue, getError, onChange } = useFormContext();

    const id = name;
    const value = getValue ? getValue(name) : otherProps.value;
    const message = getError ? getError({ name }) : errors[0]?.message;
    const hasError = !!message;

    const handleOnChange = (e) => {
        if (otherProps.onChange) {
            otherProps.onChange(e);
        } else {
            onChange(e);
        }
    };

    const selectOptions = options.map(
        ({
            id,
            [valueKey]: optionValue,
            [labelKey]: optionLabel,
            ...otherProps
        }) => (
            <option key={id} value={optionValue} {...otherProps}>
                {optionLabel}
            </option>
        )
    );

    const _wrapperClass = useMemo(() => {
        let label = `mb-3 form-floating ${
            wrapperClass ? "wrapperClass" : "form-group"
        }`;

        if (inlineLabel) {
            label += " d-flex";
        }

        return label;
    }, [wrapperClass, inlineLabel]);

    const selectClass = useMemo(() => {
        return `${inlineLabel ? "col " : ""}${
            className || `${hasError ? "is-invalid" : ""} ${classNameAppend}`
        }`;
    }, [inlineLabel, className, hasError, classNameAppend]);

    return (
        <div className={_wrapperClass}>
            <FloatingLabel id={id} label={label}>
                <Form.Select
                    className={selectClass}
                    id={name}
                    name={name}
                    ref={ref}
                    value={value}
                    disabled={disabled}
                    onChange={handleOnChange}
                >
                    {selectOptions}
                </Form.Select>
            </FloatingLabel>
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback d-block">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
Select.displayName = "Select";

export default React.forwardRef(Select);
