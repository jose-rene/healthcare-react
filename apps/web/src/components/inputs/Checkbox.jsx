import React from "react";
import { Form } from "react-bootstrap";

const Checkbox = (
    {
        name,
        label = "",
        inline = false,
        type = "text",
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        labelLeft = false,
        ...otherProps
    },
    ref
) => {
    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message;

    return (
        <div className={inline ? "form-check form-check-inline" : "form-group"}>
            {labelLeft === false && label &&
            <Form.Label htmlFor={name}>{label}</Form.Label>}
            <input
                type="checkbox"
                id={name}
                name={name}
                autoComplete={autocomplete || name}
                className={
                    className ||
                    `form-check-input ${
                        hasError ? " is-invalid" : ""
                    } ${classNameAppend}`
                }
                ref={ref}
                {...otherProps}
            />
            {labelLeft && label &&
            <Form.Label htmlFor={name}>{label}</Form.Label>}
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback mt-3">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
Checkbox.displayName = "Checkbox";

export default React.forwardRef(Checkbox);
