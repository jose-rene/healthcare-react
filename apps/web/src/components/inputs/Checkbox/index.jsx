import React from "react";
import { Form } from "react-bootstrap";

const Checkbox = (
    {
        name,
        label = "",
        inline = false,
        type = "checkbox",
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        labelLeft = false,
        hasError: _hasError = false,
        ...otherProps
    },
    ref
) => {
    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message || _hasError;

    return (
        <div className={inline ? "form-check form-check-inline" : "form-group"}>
            {labelLeft === false && label &&
            <Form.Label htmlFor={name} className="pe-2">{label}</Form.Label>}
            <input
                type={type}
                id={name}
                name={name}
                autoComplete={autocomplete || name}
                className={
                    className ||
                    `form-check-input ${
                        hasError ? " is-invalid" : ""
                    } ${classNameAppend}`
                }
                ref={ref}
                {...otherProps}
            />
            {labelLeft && label &&
            <Form.Label htmlFor={name} className="ps-2">{label}</Form.Label>}
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && message && (
                <Form.Text className="invalid-feedback mt-3">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
Checkbox.displayName = "Checkbox";

export default React.forwardRef(Checkbox);
