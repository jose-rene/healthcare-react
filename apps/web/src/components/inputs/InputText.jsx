import React from "react";
import { Form, InputGroup } from "react-bootstrap";
import Icon from "../elements/Icon";

const InputText = (
    {
        name,
        prepend = null,
        label = "",
        type = "text",
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        ...otherProps
    },
    ref
) => {
    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message;

    const getInput = () => {
        return (
            <input
                id={name}
                type={type}
                name={name}
                autoComplete={autocomplete || name}
                className={
                    className ||
                    `form-control ${
                        hasError ? " is-invalid" : ""
                    } ${classNameAppend}`
                }
                ref={ref}
                {...otherProps}
            />
        );
    };

    return (
        <div className="form-group">
            {label && <Form.Label htmlFor={name}>{label}</Form.Label>}
            {prepend ? (
                <InputGroup>
                    <InputGroup.Prepend>
                        <InputGroup.Text id="basic-addon1">
                            <Icon icon={prepend} size="lg" />
                        </InputGroup.Text>
                    </InputGroup.Prepend>{" "}
                    {getInput()}
                </InputGroup>
            ) : (
                getInput()
            )}
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback mt-3">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
InputText.displayName = "InputText";

export default React.forwardRef(InputText);
