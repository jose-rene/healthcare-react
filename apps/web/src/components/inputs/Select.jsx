import React from "react";
import { Form } from "react-bootstrap";

const Select = (
    {
        name,
        label = "",
        inlineLabel = false,
        options = [],
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        labelClass = "",
        wrapperClass = false,
        labelKey = "title",
        valueKey = "val",
        ...otherProps
    },
    ref
) => {
    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message;
    const selectOptions = options.map((option) => (
        <option key={option.id} value={option[valueKey]}>
            {option[labelKey]}
        </option>
    ));
    return (
        <div
            className={`${wrapperClass ? "wrapperClass" : "form-group"} ${
                inlineLabel ? "d-flex" : ""
            }`}
        >
            {label && (
                <Form.Label
                    htmlFor={name}
                    className={
                        (labelClass || "") + (inlineLabel ? " col pt-2" : "")
                    }
                >
                    {label}
                </Form.Label>
            )}
            <select
                id={name}
                name={name}
                autoComplete={autocomplete || name}
                className={`${inlineLabel ? "col " : ""}${
                    className ||
                    `form-control ${
                        hasError ? " is-invalid" : ""
                    } ${classNameAppend}`
                }`}
                ref={ref}
                {...otherProps}
            >
                {selectOptions}
            </select>
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback mt-3">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
Select.displayName = "Select";

export default React.forwardRef(Select);
