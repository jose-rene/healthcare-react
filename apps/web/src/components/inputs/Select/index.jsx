import React, { useMemo } from "react";
import { Form, FloatingLabel } from "react-bootstrap";

const BaseSelect = (
    {
        name,
        value,
        label = "",
        inlineLabel = false,
        options = [],
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        labelClass = "",
        wrapperClass = false,
        labelKey = "title",
        valueKey = "val",
        addEmpty = false,
        emptyLabel = "Select an option",
        onChange,
        ...otherProps
    },
    ref,
) => {
    const id = name;

    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message;
    const selectOptions = options.map(({ id, [valueKey]: optionValue, [labelKey]: optionLabel, ...otherProps }) => (
        <option key={id} value={optionValue} {...otherProps}>
            {optionLabel}
        </option>
    ));

    const _wrapperClass = useMemo(() => {
        let label = `mb-3 form-floating ${
            wrapperClass ? "wrapperClass" : "form-group"
        }`;

        if (inlineLabel) {
            label += " d-flex";
        }

        return label;
    }, [wrapperClass, inlineLabel]);

    const selectClass = useMemo(() => {
        return `${inlineLabel ? "col " : ""}${
            className || `${hasError ? "is-invalid" : ""} ${classNameAppend}`
        }`;
    }, [inlineLabel, className, hasError, classNameAppend]);

    return (
        <div className={_wrapperClass}>
            <FloatingLabel id={id} label={label}>
                <Form.Select
                    className={selectClass}
                    id={name}
                    name={name}
                    ref={ref}
                    value={value}
                    onChange={onChange}
                >
                    {selectOptions}
                </Form.Select>
            </FloatingLabel>
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback d-block">
                    {message}
                </Form.Text>
            )}
        </div>
    );
};
BaseSelect.displayName = "Select";

export default React.forwardRef(BaseSelect);
