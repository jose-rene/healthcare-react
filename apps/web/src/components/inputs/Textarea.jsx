import React, { useEffect, createRef } from "react";
import { Form, FloatingLabel } from "react-bootstrap";

const Textarea = (
    {
        name,
        label = "",
        autocomplete = false,
        className = false,
        classNameAppend = "",
        helpText = false,
        errors = {},
        ...otherProps
    },
    ref
) => {
    const { [name]: { message = false } = {} } = errors;
    const hasError = !!message;
    const _ref = ref || createRef();

    useEffect(() => {
        if (otherProps.initialValue) {
            console.log("ref", { _ref });
            _ref.current.value = otherProps.initialValue;
        }

        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return (
        <FloatingLabel label={label} className="mb-3 form-group">
            <textarea
                id={name}
                name={name}
                autoComplete={autocomplete || name}
                className={
                    className ||
                    `form-control ${
                        hasError ? " is-invalid" : ""
                    } ${classNameAppend}`
                }
                ref={_ref}
                {...otherProps}
            />
            {helpText && (
                <Form.Text className="text-muted form-text">
                    {helpText}
                </Form.Text>
            )}
            {hasError && (
                <Form.Text className="invalid-feedback mt-3">
                    {message}
                </Form.Text>
            )}
        </FloatingLabel>
    );
};
Textarea.displayName = "Textarea";

export default React.forwardRef(Textarea);
