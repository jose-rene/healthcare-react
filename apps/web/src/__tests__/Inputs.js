import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import Checkbox from "../components/inputs/Checkbox";
import Textarea from "../components/inputs/Textarea";
import InputText from "../components/inputs/InputText";
import Select from "../components/inputs/Select";

// bug with coverage https://github.com/facebook/create-react-app/issues/8689

describe("Input Components", () => {
    it("renders Text input correctly", () => {
        const input = renderer.create(<InputText />).toJSON();
        expect(input).toMatchSnapshot();
    });
    it("renders Checkbox input correctly", () => {
        const input = renderer.create(<Checkbox />).toJSON();
        expect(input).toMatchSnapshot();
    });
    it("renders Textarea input correctly", () => {
        const input = renderer.create(<Textarea />).toJSON();
        expect(input).toMatchSnapshot();
    });
    it("renders Select input correctly", () => {
        const input = renderer.create(<Select />).toJSON();
        expect(input).toMatchSnapshot();
    });
});
