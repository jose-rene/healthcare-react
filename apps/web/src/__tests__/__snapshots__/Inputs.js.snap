// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Components renders Checkbox input correctly 1`] = `
<div
  className="form-group"
>
  
  <input
    className="form-control  "
    type="checkbox"
  />
</div>
`;

exports[`Input Components renders Select input correctly 1`] = `
<div
  className=""
>
  
  <select
    className="form-control  "
  />
</div>
`;

exports[`Input Components renders Text input correctly 1`] = `
<div
  className="form-group"
>
  
  <input
    className="form-control  "
    type="text"
  />
</div>
`;

exports[`Input Components renders Textarea input correctly 1`] = `
<div
  className="form-group"
>
  
  <textarea
    className="form-control  "
    type="text"
  />
</div>
`;
